### PR TITLE
Fixed #1750: Unexpected behavior when updation of talawa admin members profile. 

### LINKED ISSUE #1750: Unexpected behavior when updation of talawa admin members profile.
**Describe the bug**

This bug report pertains to an issue encountered in the Talawa-Admin platform, where:

1. When attempting to update a member's profile, it seems to require changing the email ID.
2. Despite updating the profile, the member's actual profile remains unchanged. Instead, the user's (your) profile credentials are updated to match the member's credentials.

**Issue Link raised in Talawa Admin repository**
https://github.com/PalisadoesFoundation/talawa-admin/issues/1473

**To Reproduce**

1. Open Talawa Admin website
2. Click on your profile at the bottom left
3. Click People Tab on the sidebar
4. Press users checkbox and update a members profile

**Expected behavior**
1. When a user updates a member's profile, the member's profile should get updated even if only the name is edited.
2. After updating the member's profile, it should not affect the user's (your) profile.

**Additional details**
Modifications have been made in Talawa-Admin to address this issue, which also requires corresponding adjustments in Talawa-API.
Can i work on this issue?

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by Devesh326**
**What kind of change does this PR introduce?**

bugfix

**Issue Number:**

Fixes #1750

**Did you add tests for your changes?**

No

**Snapshots/Videos:**

[](https://github.com/PalisadoesFoundation/talawa-api/assets/76838871/d884d3c8-b20d-4663-a5ce-214485ffb631)

**Summary**

Solves the following issues:

1. When attempting to update a member's profile, it required changing the email ID.
2. Despite updating the profile, the member's actual profile remained unchanged. Instead, the user's (your) profile credentials w were updated to match the member's credentials.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by Cioppolo14** [comment]
@EshaanAgg @xoldyckk Can you review this PR?

File: src/resolvers/Mutation/updateUserProfile.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/updateUserProfile.ts:62]: why would you allow anyone to update any user's profile they want to?

File: src/typeDefs/inputs.ts
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:342]: `appLanguageCode`

File: src/resolvers/Mutation/updateUserProfile.ts
Comment (xoldyckk) [inline src/resolvers/Mutation/updateUserProfile.ts:105]: `args.data?.appLanguageCode`

File: src/typeDefs/inputs.ts
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:344]: rename `input UpdateUserInput` to `input UpdateUserProfileInput`

make following changes for mutation:-
```graphql
type UpdateUserProfilePayload {
  user: User
}

type Mutation {
  updateUserProfile(input: UpdateUserProfileInput!): UpdateUserProfilePayload
}
```

File: src/typeDefs/inputs.ts
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:343]: remove the `id` field from here, the update will only be applied for a user with `_id === context.userId`

Comment (xoldyckk) [reply src/typeDefs/inputs.ts:343]: if this mutation is to be allowed to be triggered by both normal users and superadmins, some changes have to be made:-

1. if `args.input.id` is not provided(it is `null` or `undefined`), the update will be carried out on the user with `user._id` equals to `context.userId`

1. if `args.input.id` is provided, there are two scenarios:-
    1. if it's a normal user, `context.userId` must be equal to `args.input.id`
    2. if it's a superadmin there are no restrictions, they can update any user with `user._id` equals to `args.input.id`

some things to confirm with @palisadoes :-
1. can one superadmin update other superadmin's user data?
2. is this allowed only for superadmins, or admins of organizations too? the thing to remember is that a user is not tied to any one organization, they can be members/admins of many organizations simultaneously, **_this functionality doesn't exist in talawa-api right now because of existence of `userType === "ADMIN"` field on a user object, when in actuality the relationship between a user and an organization should exist on a junction between them, joining table in sql it is called_**

Comment (palisadoes) [reply src/typeDefs/inputs.ts:343]: 1. @xoldyckk is correct on the userType to organization relationship. This should be fixed in this issue.
    1. https://github.com/PalisadoesFoundation/talawa-api/issues/1711
1. Senior usertypes should be able to edit or demote users of junior rank. They should be able to promote junior ranks to no more than their rank. Super Admins will need privileges to do this for everyone. I'm broadly basing this on Google's guidelines for their apps.
    1. https://apps.google.com/supportwidget/articlehome?hl=en&article_url=https%3A%2F%2Fsupport.google.com%2Fa%2Fanswer%2F1219251%3Fhl%3Den&assistant_event=welcome&assistant_id=usermasterbot&product_context=1219251&product_name=UnuFlow&trigger_context=a

File: README.md
Comment (SiddheshKukade) [inline README.md:31]: This is not required

File: src/resolvers/Mutation/updateUserProfile.ts
Comment (SiddheshKukade) [reply src/resolvers/Mutation/updateUserProfile.ts:62]: I think this api call would be done by the admin only.

**Comment by akhilender-bongirwar** [comment]
@Devesh326, please make the reviewed changes ASAP? All the PRs in the "talawa-admin" repo are failing the graphql- introspection workflow.

**Comment by Devesh326** [comment]
Sorry for the delayed response. I was travelling, so I wasn't able to make changes to the PR. Will update and resolve the requested changes ASAP.
Apologies again for the delay.

File: src/typeDefs/inputs.ts
Comment (Devesh326) [reply src/typeDefs/inputs.ts:343]: Hey @xoldyckk,

I included the `id` in the mutation because during the API call to edit a specific user, in the backend `context.userId` only contained the `id` of the logged-in user. As a result, the profile being updated was that of the current user rather than the intended one. To address this, I passed the `id` of the user that requires editing/updating in the API call.

Following this logic, if `args.input.id` is `null`, it indicates the root user; otherwise, it's the `id` of the user whose profile needs updating.

File: src/typeDefs/inputs.ts
Comment (Devesh326) [reply src/typeDefs/inputs.ts:342]: Hey, 

Regarding the variable name; that the frontend uses `applangcode`, so should I change it there also, otherwise it would give errors?

**Comment by palisadoes** [comment]
*NOTE* Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1840
1. It reformatted over 400 files. This will put your PR at risk of extensive merge conflicts.    
1. I suggest that before you merge your code to the latest upstream, or manually fix these merge conflicts that you:
    1. upgrade your `prettier` in your local branch to the same version
    1. run `prettier`
    
This will help to reduce the number of future merge conflicts for your PR.

File: src/typeDefs/inputs.ts
Comment (SiddheshKukade) [reply src/typeDefs/inputs.ts:376]: Yes please do that @Devesh326

File: src/typeDefs/inputs.ts
Comment (SiddheshKukade) [reply src/typeDefs/inputs.ts:377]: ok. then it looks fine.

File: src/typeDefs/inputs.ts
Comment (SiddheshKukade) [reply src/typeDefs/inputs.ts:378]: Please renamed to suit the conventions @Devesh326

**Review by SiddheshKukade — CHANGES_REQUESTED** [review]
Please fix the small changes in comments. Then it's ready to be merged.

File: src/typeDefs/inputs.ts
Comment (xoldyckk) [reply src/typeDefs/inputs.ts:376]: @Devesh326 stop using abbreviations, it helps no one reading abbreviations to make sense of things, most of the time your code should be self documenting

in that light actually make it `applicationLanguageCode`

if frontend makes wrong assumptions about things, it doesn't mean it will be enforced on the backend design, if frontend wants to make changes it has to first open an issue/feature request

File: src/typeDefs/inputs.ts
Comment (xoldyckk) [reply src/typeDefs/inputs.ts:377]: @Devesh326 take some time and contemplate about what i said previously, the current logic as it stands allows a normal user to update user fields of any other user, they just need to pass in the other user's `id` in the input, frontend application can't prevent normal users from making these requests

understand the authentication/authorization flows, self review your code and possible edge cases, only then ask for reviews from other contributers

File: src/resolvers/Mutation/updateUserProfile.ts
Comment (xoldyckk) [reply src/resolvers/Mutation/updateUserProfile.ts:62]: this restriction isn't put in place, right now it allows any user to call this mutation

File: src/resolvers/Mutation/updateUserProfile.ts
Comment (xoldyckk) [reply src/resolvers/Mutation/updateUserProfile.ts:105]: this hasn't been resolved still

**Comment by akhilender-bongirwar** [comment]
@Devesh326, are you still actively working on this? The completion of this pull request is crucial, as it is causing failures in the workflow checks for all other pull requests in the admin. Please prioritize and resolve this issue as soon as possible.

**Comment by Devesh326** [comment]
> Please fix the small changes in comments. Then it's ready to be merged.

yeah okay, fixing the necessary changes

File: src/typeDefs/inputs.ts
Comment (Devesh326) [reply src/typeDefs/inputs.ts:391]: Done

File: src/resolvers/Mutation/updateUserProfile.ts
Comment (palisadoes) [reply src/resolvers/Mutation/updateUserProfile.ts:62]: This needs to be resolved. Only the profile's user, Admins and Super Admins must be able to edit profile settings.

**Review by palisadoes — CHANGES_REQUESTED** [review]

File: src/typeDefs/inputs.ts
Comment (Devesh326) [reply src/typeDefs/inputs.ts:390]: Yes noted.

**Comment by xoldyckk** [comment]
@palisadoes if one user could join multiple organizations under the same user profile, why would admin of any one organization be able to edit their user profile? that profile change would be reflected in all other organizations for that particular user. shouldn't this only be allowed to superadmins?

**Comment by palisadoes** [comment]
The design philosophy of making users become members of any organization means that operationally if a member asks an admin to update their information, such as a phone number (or possibly in the future credit card information) the admin won’t be able to do it for them.

At some future date we’ll need to think of ways of restricting this to superadmins on a per organization basis.

**Comment by palisadoes** [comment]
Please try to get the code coverage for the patch closer to that of the repository

**Comment by palisadoes** [comment]
We would like to merge this PR, however you haven’t addressed the reviewer questions. Please address each one and mark them as resolved when the work is completed. Only then will they be able to approve your work so that it can be merged into the code base.

Please also try to get the code coverage for the patch higher. 80% coverage is much lower than the 98% enjoyed by the project

**Comment by palisadoes** [comment]
This is an update on the PR merging freeze:

1. In the next few hours we will be merging the `develop-userTypeFix` branch into the `develop` branch. 
1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. This merge will cause some conflicts in your PR.

We decided to do this at the beginning of the weekend to give us all time to adjust PR code and create bug fixes that may arise.

Update your code at or after midnight GMT on the morning of March 23, 2024. (5:30am IST). 

If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by palisadoes** [comment]
## The PR merging freeze is lifted.

1. We are working on bug fixes that may arise. These should be few and are being tracked here:
    1. https://github.com/orgs/PalisadoesFoundation/projects/24
1. Please update your PRs and local repos. Fix any new merge conflicts that may have occurred.
    
Background:

1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by Cioppolo14** [comment]
@Devesh326 Please fix the conflicting files.

**Comment by palisadoes** [comment]
I'm going to close this as abandoned. There have been no commits for over a month.

**Pull request closed without merging**
